src_c/transform.c

```diff
@@ -570,6 +609,110 @@ scale_to(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj, int width,
     return retsurf;
 }
 
+SDL_Surface *
+grayscale(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj)
+{
+    SDL_Surface *src = pgSurface_AsSurface(srcobj);
+    SDL_Surface *newsurf;
+
+    if (!dstobj) {
+        newsurf = newsurf_fromsurf(src, srcobj->surf->w, srcobj->surf->h);
+        if (!newsurf)
+            return NULL;
+    }
+    else {
+        newsurf = pgSurface_AsSurface(dstobj);
+    }
+
+    if (newsurf->w != src->w || newsurf->h != src->h) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Destination surface must be the same size as source surface."));
+    }
+
+    if (src->format->BytesPerPixel != newsurf->format->BytesPerPixel) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Source and destination surfaces need the same format."));
+    }
+
+    Uint8 *pixel_8 = NULL;
+    Uint16 *pixel_16 = NULL;
+    Uint32 *pixel_32 = NULL;
+
+    if (src->format->BitsPerPixel == 8) {
+        pixel_8 = (Uint8 *)src->pixels;
+    }
+    else if (src->format->BitsPerPixel == 16) {
+        pixel_16 = (Uint16 *)src->pixels;
+    }
+    else if (src->format->BitsPerPixel == 32) {
+        pixel_32 = (Uint32 *)src->pixels;
+    }
+
+    int x, y;
+    for (y = 0; y < newsurf->h; y++) {
+        for (x = 0; x < newsurf->w; x++) {
+            int i = y * newsurf->pitch / newsurf->format->BytesPerPixel + x;
+
+            Uint8 r, g, b, a;
+            if (pixel_8) {
+                SDL_GetRGBA(pixel_8[i], newsurf->format, &r, &g, &b, &a);
+            }
+            else if (pixel_16) {
+                SDL_GetRGBA(pixel_16[i], newsurf->format, &r, &g, &b, &a);
+            }
+            else if (pixel_32) {
+                SDL_GetRGBA(pixel_32[i], newsurf->format, &r, &g, &b, &a);
+            }
+            else {
+                SDL_GetRGBA(pixel_32[i], newsurf->format, &r, &g, &b, &a);
```

Comment:
It seems like it should be using the src surface's format, because that's what it is reading from.